### PR TITLE
mpl2: post-annealing step to push macros to boundaries

### PR DESCRIPTION
**Pull request by AcKoucher**
[#1930](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1930) showed that we actually don't allow different x & y coordinates for macro placement halos in mpl2, so [#1933](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/1933) was opened.

When I was testing the distinct x & y halos for mock-array's test case used in #4085, two problems came up:
- The macros' locations inside the array were wrong (should be fixed by #4976)
- The array was getting split in two as mpl2 was trying to push the macros to the boundaries by a post-annealing function.

This PR is to change the alignHardMacroGlobal function by a new step in mpl2 that happens before we improve orientation. Basically, we push macros that are close enough to the boundaries if:
1. They don't overlap with blockages generated for Pin Access
2. They don't overlap with other macros
3. Pushing won't destroy the shape of an array of macros.

### CODE REVIEW

**Comment by AcKoucher** [comment]
It looks like there's something going on with clang-tidy.

**Comment by maliberty** [comment]
secure ci?

File: src/mpl2/src/hier_rtlmp.cpp [hunk 1]
```diff
@@ -6638,4 +6241,235 @@ void HierRTLMP::setDebugShowBundledNets(bool show_bundled_nets)
...
+  std::vector<HardMacro*> hard_macros = macro_cluster->getHardMacros();
+
+  int min_dx_to_core = std::numeric_limits<int>::max();
+  int min_dy_to_core = std::numeric_limits<int>::max();
+
+  for (HardMacro* hard_macro : hard_macros) {
+    hard_macros_.push_back(hard_macro);
+
+    min_dx_to_core = std::min(min_dy_to_core, hard_macro->getWidthDBU());
+    min_dy_to_core = std::min(min_dy_to_core, hard_macro->getHeightDBU());
```
Comment (maliberty) [inline src/mpl2/src/hier_rtlmp.cpp:6329]: The variable name seems to suggest a distance but the value is w/h.  Please explain your thinking.

File: src/mpl2/src/hier_rtlmp.cpp [hunk 2]
```diff
@@ -6638,4 +6241,235 @@ void HierRTLMP::setDebugShowBundledNets(bool show_bundled_nets)
...
+        macro_cluster, vertical_move_allowed, horizontal_move_allowed);
+
+    pushMacroClusterToCoreBoundaries(macro_cluster, boundaries_distance);
+  }
+}
+
+std::map<Boundary, int> Pusher::getDistanceToCloseBoundaries(
+    Cluster* macro_cluster,
+    bool vertical_move_allowed,
+    bool horizontal_move_allowed)
```
Comment (maliberty) [inline src/mpl2/src/hier_rtlmp.cpp:6316]: const

File: src/mpl2/src/hier_rtlmp.cpp (hunk 1 above)
Comment (AcKoucher) [reply src/mpl2/src/hier_rtlmp.cpp:6329]: The names should just be w/h as the value, those names weren't a good choice. I also realized that we don't need to iterate the HardMacros within the macro cluster and make all this std::min stuff as we only group macros of the same size (I'll comment this on the code).

**Comment by AcKoucher** [comment]
I started a secure-ci

**Comment by maliberty** [comment]
@osamahammad21 this comes with one new drc in ca53 but it seems to be just a case of perturbing things happens to trigger a routing issue.  I don't see a problem in the macro placement so I'll ask you to look at it but don't want to hold this PR for it.

**Pull request merged**